Source/JavaScriptCore/runtime/WaiterListManager.h

```diff
@@ -0,0 +1,229 @@
+    RefPtr<RunLoop::DispatchTimer> m_timer { nullptr };
+    Condition m_condition;
+    bool m_isAsync { false };
+};
+
+class WaiterList : public ThreadSafeRefCounted<WaiterList> {
+    WTF_MAKE_FAST_ALLOCATED;
+
+public:
+    ~WaiterList()
+    {
+        removeIf([](Waiter*) {
+            return true;
+        });
+    }
+
+    void addLast(const AbstractLocker&, Waiter* waiter)
+    {
+        m_waiters.append(waiter);
+        // Since SyncWaiter will be allocated on stack, we only need to ref AsyncWaiter.
+        if (waiter->isAsync())
+            waiter->ref();
+        m_size++;
+    }
+
+    Ref<Waiter> takeFirst(const AbstractLocker&)
+    {
+        return take(&*m_waiters.begin());
+    }
+
+    Ref<Waiter> take(const AbstractLocker&, Waiter* waiter)
+    {
+        return take(waiter);
+    }
+
+    template<typename Functor>
+    void removeIf(const AbstractLocker&, const Functor& functor)
+    {
+        removeIf(functor);
+    }
+
+    unsigned size()
+    {
+        return m_size;
+    }
+
+    Lock lock;
+
+private:
+    template<typename Functor>
+    void removeIf(const Functor& functor)
+    {
+        m_waiters.forEach([&](Waiter* waiter) {
+            if (functor(waiter))
+                removeWithUpdate(waiter);
+        });
+    }
+
+    Ref<Waiter> take(Waiter* waiter)
+    {
+        // `take` function is used to consume a waiter (either notify or timeout).
+        // First, we need make sure it's not removed from the list.
+        ASSERT(waiter->vm() || waiter->getTicket());
+        Ref<Waiter> protectedWaiter = Ref { *waiter };
+        removeWithUpdate(waiter);
+        return protectedWaiter;
+    }
+
+    void removeWithUpdate(Waiter* waiter)
+    {
+        m_waiters.remove(waiter);
+        // Since SyncWaiter will be allocated on stack, we only need to deref AsyncWaiter.
+        if (waiter->isAsync())
+            waiter->deref();
+        m_size--;
+    }
+
+    unsigned m_size { 0 };
+    SentinelLinkedList<Waiter, BasicRawSentinelNode<Waiter>> m_waiters;
+};
+
+class WaiterListManager {
+    WTF_MAKE_FAST_ALLOCATED;
+
+public:
+    static WaiterListManager& singleton();
+
+    JS_EXPORT_PRIVATE JSValue wait(JSGlobalObject*, VM&, void* ptr, AtomicsWaitValidation, Seconds timeout, AtomicsWaitType);
+
+    void addAsyncWaiter(void* ptr, JSPromise*, Seconds timeout);
+
+    enum class ResolveResult : uint8_t { Ok, Timeout };
+    unsigned notifyWaiter(void* ptr, unsigned count);
+
+    size_t waiterListSize(void* ptr);
+
+    void unregisterVM(VM*);
+
+    void unregisterSharedArrayBuffer(uint8_t* arrayPtr, size_t);
+
+private:
+    void notifyWaiterImpl(const AbstractLocker&, Ref<Waiter>&&, const ResolveResult);
+
+    void timeoutAsyncWaiter(void* ptr, Waiter* target);
+
+    void cancelAsyncWaiter(Waiter*);
+
+    Ref<WaiterList> findOrCreateList(void* ptr);
+
+    RefPtr<WaiterList> findList(void* ptr);
+
+    Lock m_waiterListsLock;
+    HashMap<void*, std::unique_ptr<WaiterList>> m_waiterLists;
```

Comment:
`std::unique_ptr` is wrong. It is ref-counted. Then you should use `Ref<WaiterList>`